enhancements/machine-api/out-of-tree-provider-support.md

```diff
@@ -0,0 +1,553 @@
+
+The plan for migration is the following:
+
+1) `kubelet` and `kube-controller-manager`.
+    1) Set the `CloudControllerOwner` condition on the `ClusterOperator` with `status=True`.
+    2) Disable cloud loops in its operands.
+    3) Wait for the new resource version rollout to complete and become stable.
+    4) Indicate the completion of the migration by setting `status=False` on the `CloudControllerOwner` condition.
+2) `cloud-controller-manager`
+    1) Set the `CloudControllerOwner` condition on the `ClusterOperator` with `status=False`.
+    2) Wait for preceding operators to have the `CloudControllerOwner` condition with `status=False`.
+    3) Provision out-of-tree resources.
+    4) Indicate the ownership transfer by setting `status=True` on the `CloudControllerOwner` condition.
+    3) Wait for the resources rollout to finish.
+
+Once all resources are up and running, the migration is completed.
+
+*Note:`kube-apiserver` migration is not important, as it currently depends on `cloud.Interface` only in establishing SSH node tunneling functionality unused in `OpenShift` and [deprecated](https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-apiserver/app/options/options.go#L197-L205) since 1.9*
+
+##### Migration rollback (disaster recovery)
+
+Migration rollback will be done by switching operator places. `cloud-controller-manager` will become an operator in 1st step and `kubelet` together with `kube-controller-manager` will be in the 2d.
+
+##### Migration Lifecycle
+
+Before it is decided to make out-of-tree a default selection for the release, the 1st step will be initiated by the `FeatureGate` resource described above. In the designated release for switch from in-tree for a particular platform, the 1st step will be a part of the default upgrade procedure.
+
+Once the provider is moved to out-of-tree, the migration mechanism will be disabled and the helper `CloudControllerOwner` condition will be removed. When all existing in-tree providers are moved to out-of-tree, the implementation will be safe to remove.
+
+#### Bootstrap changes
+
+Once an out-of-tree provider is released (GA), the `CCM` will be created as a static pod on the bootstrap node, to ensure swift removal of the `node.cloudprovider.kubernetes.io/uninitialized: true` taint from any newly created `Nodes`. Later stages, including cluster upgrades will be managed by an operator, which will ensure stability of the configuration, and will run the `CCM` in a `Deployment`.
+Initial usage of the static pod is justified by the need for `CCM` to initialise `Nodes` before the `kube-scheduler` is able to schedule regular workloads (eg the operator for `CCM`).
+
+The order will be the following:
+
+1) Start `etcd` and `kube-api-server`.
+2) Run the `CCM`, which will be provisioned as a static pod.
+3) Start the `kube-scheduler`.
+
+`CCCMO` will provide a way to generate initial manifests that allow deployment of `CCM` as a static pod on the bootstrap machine.
```

Comment:
Based on Stefan's comment
```suggestion
`CCCMO` will provide a render command to generate initial manifests that allow deployment of `CCM` as a static pod on the bootstrap machine. This will follow the style of other operators (eg `KCMO`).
```